lib/io/hardware_gpio.hpp

```diff
@@ -11,23 +12,53 @@ class HardwareGpioReader : public IGpioReader {
   virtual std::optional<core::DigitalSignal> read();
 
  private:
-  HardwareGpioReader();
+  HardwareGpioReader(uint8_t pin, volatile unsigned int *read): pinMAP(pin), gpio_readAddr(read){};
+
+  const uint8_t pinMAP;
+  volatile unsigned int *gpio_readAddr;
+
   friend class HardwareGpio;
 };
 
+
+
+
+
+
+
 class HardwareGpioWriter : public IGpioWriter {
  public:
   virtual GpioWriteResult write(const core::DigitalSignal state);
 
  private:
-  HardwareGpioWriter(const uint8_t pin);
+  HardwareGpioWriter(const uint8_t pin,  volatile unsigned int *set, volatile unsigned int *clear): pinMAP(pin), gpio_setAddr(set), gpio_clearAddr(clear){};
+  
+  const uint8_t pinMAP;
+  volatile unsigned int *gpio_setAddr;
+  volatile unsigned int *gpio_clearAddr;
+
   friend class HardwareGpio;
 };
 
 /**
  * Hardware GPIO interface, requires physical GPIO pins to be present. This should only
  * be instantiated at the top level and then provided to users through the IGpio interface.
  */
+
+//GPIO  Start Addr   End Addr
+//GPIO0 0x44E0_7000  0x44E0_7FFF
+//GPIO1 0x4804_C000  0x4804_CFFF
+//GPIO2 0x481A_C000  0x481A_CFFF
+//GPIO3 0x481A_E000  0x481A_EFFF
+
+//GPIO_DATAIN (READ) 0x138h
+//GPIO_DATAOUT 0x13c
+//SET 0x194
+//CLEAR 0x190
+//  const uint8_t bank   = pin_ / 32;  // offset: GPIO_0,1,2,3
+//  const uint8_t pin_id = pin_ % 32;
+
```

Comment:
commented code?

Reply:
Referenced information for getting pin addresses. Can delete but anybody later on reading might find it useful.

Reply:
If it's stuff you need to actually use in the code itself, don't keep it commented and just comment on the method that needs it desribing briefly how to use it.